src/hotspot/share/gc/g1/heapRegionSet.cpp

```diff
@@ -231,36 +231,37 @@ void FreeRegionList::remove_starting_at(HeapRegion* first, uint num_regions) {
   check_mt_safety();
   assert_free_region_list(num_regions >= 1, "pre-condition");
   assert_free_region_list(!is_empty(), "pre-condition");
+  assert_free_region_list(length() >= num_regions, "pre-condition");
 
   verify_optional();
   DEBUG_ONLY(uint old_length = length();)
 
+  // prev points to the node right before first or null when first == _head
+  HeapRegion* const prev = first->prev();
+  if (prev == NULL) {
+    assert_free_region_list(_head == first, "invariant");
+  } else {
+    assert_free_region_list(_head != first, "invariant");
+  }
+  // next points to the node right after first or null when first == _tail
+  HeapRegion* next = first->next();
+  if (next == NULL) {
+    assert_free_region_list(_tail == first, "invariant");
+  } else {
+    assert_free_region_list(_tail != first, "invariant");
+  }
```

Comment:
Since these checks no longer does anything other than assertions I think it would be nice to hide it in a helper that for production builds will do nothing using `NOT_DEBUG_RETURN`.

Reply:
Thanks for revewing, just changed as you suggested.

Reply:
You are correct that when executing it will have the same effect, but I'm more concerned about the readability of the code in this case. So please move it to a separate function, something like `verify_region_to_remove()`. Looking a bit closer at the code I think this function should be able to be called from each round of the loop as well, I mean that's what was done before and by doing so we could get the loop a bit cleaner as well. Or have I missed anything? In that case you would not need to call it before the loop, but just inside it.

Reply:
Thank you for reviewing and clarification. Got your point, I have changed as you suggested, would you mind to have another look at it?